compiler/hash-typecheck/src/operations/args.rs

```diff
@@ -0,0 +1,268 @@
+    },
+    tc::Tc,
+    traits::{OperationsOnNode, RecursiveOperationsOnNode},
+    utils::matching::MatchResult,
+};
+
+impl<E: TcEnv> RecursiveOperationsOnNode<ArgsId> for Tc<'_, E> {
+    type TyNode = ParamsId;
+    type RecursiveArg = ArgsId;
+
+    fn check_node_rec<T, F: FnMut(Self::RecursiveArg) -> TcResult<T>>(
+        &self,
+        args: ArgsId,
+        annotation_params: Self::TyNode,
+        mut in_arg_scope: F,
+    ) -> TcResult<T> {
+        self.register_new_atom(args, annotation_params);
+        let reordered_args_id = validate_and_reorder_args_against_params(args, annotation_params)?;
+
+        let result = self.check_some_args(
+            reordered_args_id.iter(),
+            annotation_params,
+            |arg, param_ty| {
+                let arg = arg.value();
+                self.check_node(arg.value, param_ty)?;
+                Ok(())
+            },
+            |arg| {
+                let arg = arg.value();
+                Some(arg.value)
+            },
+            || in_arg_scope(reordered_args_id),
+        )?;
+
+        Ok(result)
+    }
+
+    fn try_normalise_node_rec(&self, args_id: ArgsId) -> NormaliseResult<ControlFlow<ArgsId>> {
+        let args = args_id.value();
+        let st = NormalisationState::new();
+
+        let evaluated_arg_data = args
+            .value()
+            .into_iter()
+            .map(|arg| -> Result<_, NormaliseSignal> {
+                Ok(Node::at(
+                    Arg {
+                        target: arg.target,
+                        value: self.normalise_nested_node_and_record(arg.value, &st)?,
+                    },
+                    arg.origin,
+                ))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let new_node =
+            Node::create_at(Node::<Arg>::seq(evaluated_arg_data), args_id.origin().computed());
+
+        normalised_if(|| new_node, &st)
+    }
+
+    fn unify_nodes_rec<T, F: FnMut(Self::RecursiveArg) -> TcResult<T>>(
+        &self,
+        src_id: ArgsId,
+        target_id: ArgsId,
+        mut f: F,
+    ) -> TcResult<T> {
+        if src_id.len() != target_id.len() {
+            return Err(TcError::DifferentParamOrArgLengths {
+                a: src_id.into(),
+                b: target_id.into(),
+            });
+        }
+        for (src_arg_id, target_arg_id) in src_id.iter().zip(target_id.iter()) {
+            let src_arg = src_arg_id.value();
+            let target_arg = target_arg_id.value();
+            self.unify_nodes(src_arg.value, target_arg.value)?;
+        }
+        f(src_id)
+    }
+}
+
+impl<E: TcEnv> Tc<'_, E> {
+    pub fn get_binds_in_pat_args(&self, pat_args: PatArgsId) -> HashSet<SymbolId> {
+        let mut binds = HashSet::new();
+        self.visitor().visit(pat_args, &mut |atom| {
+            if let Atom::Pat(pat_id) = atom {
+                match *pat_id.value() {
+                    Pat::Binding(var) => {
```

Comment:
even for `_` patterns?

Reply:
Yes, this is needed so that the unifier can make progress sometimes. For example `y: a ~ b; match y { Refl(_) => Q }` needs to substitute the unnamed symbol `_` for `a` and for `b` so that equality theorems can be proven.

Reply:
But so called “rvalues” can’t reference a `_` and have it work?

Reply:
No because name resolution happens before typechecking and pattern `_` is not treated as an identifier during resolution.